### PR TITLE
Add methods to get window props

### PR DESCRIPTION
**Pull request by chicoxyzzy**

### CODE REVIEW

File: src/desktop/main.cc [hunk 1]
```diff
@@ -610,6 +610,59 @@ MAIN {
...
+      return;
+    }
+
+    if (message.name == "window") {
+      const auto index = message.index;
+      auto window = windowFactory.getWindow(index);
+      if (window) {
+        const auto size = window->getSize();
+        const auto title = window->getTitle();
+        SSC::String value(
```
Comment (jwerle) [inline src/desktop/main.cc:642]: this is a good candidate for `SSC::JSON`:

```c++
JSON::Object value = JSON::Object::Entries {
  {"index", index},
  {"title", title},
  {"width", size.width},
  {"height", size.height}
};
```

**Comment by chicoxyzzy** [comment]
`getTitle` and `getSize` only implemented for macOS

File: src/desktop/main.cc [hunk 2]
```diff
@@ -610,13 +613,140 @@ MAIN {
...
+    if (message.name == "getWindows") {
+      const auto index = message.index;
+      const auto showTitle = message.get("title") == "true" ? true : false;
+      const auto showSize = message.get("size") == "true" ? true : false;
+      const auto showStatus = message.get("status") == "true" ? true : false;
+      const auto window = windowManager.getWindow(index);
+      StringStream ss;
+      bool first = true;
+      ss << "[";
+      for (auto windowWithMetadata : windowManager.windows) {
```
Comment (jwerle) [inline src/desktop/main.cc:625]: maybe we can add a `json()` method to the `ManagedWindow` class that returns a `JSON::Object` and a `json()` to `WindowManager` that returns an `JSON::Array` of those objects

Comment (chicoxyzzy) [reply src/desktop/main.cc:625]: that's a good idea!

Comment (chicoxyzzy) [reply src/desktop/main.cc:625]: @jwerle how can I implement these conditions using `JSON::Object`?

```js
            if (showTitle) {
               const auto title = w->getTitle();
               ss << "\"title\":\"" << title << "\",";
             }
             if (showSize) {
               const auto size = w->getSize();
               ss << "\"width\":" << size.width << ",";
               ss << "\"height\":" << size.height << ",";
             }
             if (showStatus) {
               const auto status = windowWithMetadata->status;
               ss << "\"status\":" << status << ",";
             }
             ss << "\"index\":" << windowWithMetadata->opts.index;
             ss << "}";
```

Comment (chicoxyzzy) [reply src/desktop/main.cc:625]: aha, Object::Entries is `std::map`!

**Comment by chicoxyzzy** [comment]
@jwerle @heapwolf I think it's ready. I can't run apps on Linux and Windows, so I need help with `getTitle` and `getSize` on these OSs

File: src/desktop/main.cc (hunk 2 above)
Comment (chicoxyzzy) [reply src/desktop/main.cc]: done

File: src/window/apple.mm [hunk 3]
```diff
@@ -732,6 +732,10 @@ - (NSString*) filePromiseProvider: (NSFilePromiseProvider*)filePromiseProvider f
     }
   }
 
+  SSC::String Window::getTitle () {
+    return SSC::String([this->window.title UTF8String]);
+  }
+
   void Window::setTitle (const SSC::String& seq, const SSC::String& value) {
```
Comment (jwerle) [inline src/window/apple.mm:739]: We could also store the title on the `Window` or `ManagedWindow` class instance too

File: src/window/linux.cc [hunk 4]
```diff
@@ -675,6 +675,10 @@ namespace SSC {
     }
   }
 
+  SSC::String Window::getTitle () {
+    // TODO: implement
```
Comment (jwerle) [inline src/window/linux.cc:679]: `return String(gtk_window_get_title(GTK_WINDOW(window)));`

File: src/window/linux.cc [hunk 5]
```diff
@@ -744,6 +748,10 @@ namespace SSC {
     gtk_dialog_run(GTK_DIALOG(dialog));
   }
 
+  ScreenSize Window::getSize () {
+    // TODO: implement
```
Comment (jwerle) [inline src/window/linux.cc:752]:
```c
Screen size size = {0};
gtk_window_get_size(GTK_WINDOW(window), &size.width, &size.height);
return size;
```

File: src/window/win.cc [hunk 6]
```diff
@@ -1036,6 +1036,10 @@ namespace SSC {
     });
   }
 
+  SSC::string Window::getTitle () {
+    // TODO: implement.
```
Comment (jwerle) [inline src/window/win.cc:1040]: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getwindowtexta

Comment (jwerle) [reply src/window/win.cc:1040]:
```c++
char title[1024] = {0};
auto size = GetWindowText(window, title, 1024);
return String(title, size);
```

**Review by jwerle — APPROVED** [review]
LGTM. The missing implementations are pretty straightforward. We can follow up in a subsequent PR if you don't want to add them here. I suggest we track the title of the window for fast read access instead of calling these platform APIs each time

File: src/window/apple.mm (hunk 3 above)
Comment (heapwolf) [reply src/window/apple.mm:739]: is there an opportunity for it to become out of sync though if we do that? this isn't in the performance path so I'd say its fine.

**Pull request merged**